0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+
+foreach s: Suit {
+  ...
+}
+
+foreach b1: bool, b2: bool {
+  expect TestMyFunctionWithOptions(b1, b2) == true;
+}
+
+type SmallIndex = x: nat | 0 <= x < 8
+
+foreach i: SmallIndex {
+  ...
+}
+```
+
+There is no builtin support for automatically tracking the enumerated values so far, as there is for
+`iterator` values (see https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-iterator-types). 
+It is straightforward to track these values manually, however:
+
+```dafny
+ghost var xs := [];
+foreach x | x in s {
+  ...
+  xs := xs + [x];
+}
+```
+
+It is even possible to attach this ghost state directly to the sequence with a helper function:
+
+```dafny
+foreach x, xs | (x, xs) in WithPrefixes(c) {
+  ...
+}
+```
+
+Where the signature of `WithPrefixes` is something like:
+
+```dafny
+function method WithPrefixes<T>(s: seq<T>): seq<(T, ghost seq<T>)> {
+  ...
+}
+```
+
+Similarly, a helper function can be provided to maintain a running index of the enumerated values:
+
+```dafny
+foreach x, i | (x, i) in WithIndexes(s) {
+  ...
+}
+
+// Alternatively, if s is a sequence:
+foreach x, i | 0 <= i < |s| && s[i] == x {
+  ...
+}
+```
+
+## Sequence comprehensions
+
+A sequence comprehension has identical syntax to a set comprehension, except that it begins with
+`seq` rather than `set`. Therefore its syntax is:
+
+```
+SeqComprehension ::=
+  "seq" 
+  Ident [":" Type] { "," Ident [":" Type] } 
+  "|" Expression
+  [ :: Expression ]
+```
+
+Sequence comprehensions are the expression analogues of `foreach` loops, just as
+`forall` expressions are the analogues of `forall` statements. The value of any given
+`seq <BoundVariables> | <Range> :: <Term>` expression is specified by the following `foreach` loop:
+
+```dafny
+var result := [];
+foreach <BoundVariables> | <Range> {
+  result := result + [<Term>];
+}
+```
+
+Sequence comprehensions are likely to be easier to work with than `foreach` loops, and Dafny users should be
+encouraged to use them instead where possible. Many high-order operations on sequences can be expressed directly
+using sequence comprehensions:
+
+```dafny
+// Filtering to optional values
+var s: seq<Option<T>> := ...;
+var filtered: seq<T> := seq x | x in s && x.Some? :: x.value;
+
+// Zipping two lists together
+var a: seq<T> := ...;
+var b: seq<T> := ...;
+assert |a| == |b|;
+var zipped := seq i | 0 <= i < |a| :: (a[i], b[i])
+
+// Map and then flatten (a.k.a. "FlatMap")
+var f: T -> seq<T> := ...;
+var c: seq<T> := ...;
+var flatMapped: seq<T> := seq x, y | x in c && y in f(x) :: y;
+```
+
+Note that the existing `seq(size, i => i + 1)` syntax is inconsistently-named in the Dafny reference manual, 
+but we refer to it here as a "sequence construction" expression, to disambiguate it from the proposed sequence comprehension feature.
+Sequence comprehensions are strictly more expressive, as any construction `seq(size, f)` can be rewritten as 
+`seq i | 0 <= i < size :: f(i)`.
```

Comment:
And we could eventually move `seq(size, f)` to a definition in the standard library, perhaps?

Reply:
Most likely - @RustanLeino was willing to fully deprecate the old version as long as we handled the equivalent sequence comprehension equally well in verification. Having an equivalent standard library version would make dropping support for it easier!